ngraph_bridge/ngraph_builder.cc

```diff
@@ -373,6 +373,82 @@ Builder::PerformNgBroadcast(const string& prov_tag,
+  ng::AxisSet axes{};
+  for (auto i = 0; i < sizeof(int) * 8; ++i) {
+    if ((unsigned char)(mask >> i & 0x01) == 1) {
+      axes.emplace(i);
+    }
+  }
+  return axes;
+};
+
+struct strided_slice_mask_attrs {
+  int begin;
+  int end;
+  int new_axis;
+  int shrink_axis;
+  int ellipsis;
+};
+
+static Status GetStridedSliceAttrs(const Node* op,
+                                   strided_slice_mask_attrs& attrs) {
+  TF_RETURN_IF_ERROR(GetNodeAttr(op->attrs(), "begin_mask", &attrs.begin));
+  TF_RETURN_IF_ERROR(GetNodeAttr(op->attrs(), "end_mask", &attrs.end));
+  TF_RETURN_IF_ERROR(
+      GetNodeAttr(op->attrs(), "new_axis_mask", &attrs.new_axis));
+  TF_RETURN_IF_ERROR(
+      GetNodeAttr(op->attrs(), "shrink_axis_mask", &attrs.shrink_axis));
+  TF_RETURN_IF_ERROR(
+      GetNodeAttr(op->attrs(), "ellipsis_mask", &attrs.ellipsis));
+
+  NGRAPH_VLOG(5) << "Strieded Slice Mask Attributes: "
+                 << "  begin mask: " << attrs.begin
+                 << "  end mask: " << attrs.end
+                 << "  new axis mask: " << attrs.new_axis
+                 << "  shrink axis mask: " << attrs.shrink_axis
+                 << "  ellipsis mask: " << attrs.ellipsis;
+
+  return Status::OK();
+}
+
+static Status GetSlicePlan(const ng::Shape& shape,
+                           const std::vector<int64_t>& begin,
+                           const std::vector<int64_t>& end,
+                           const std::vector<int64_t>& stride,
+                           const strided_slice_mask_attrs& mask_attrs,
+                           ng::SlicePlan& slice_plan) {
+  slice_plan = ng::make_slice_plan(
+      shape, begin, end, stride, ConvertMaskToAxes(mask_attrs.begin),
+      ConvertMaskToAxes(mask_attrs.end), ConvertMaskToAxes(mask_attrs.new_axis),
+      ConvertMaskToAxes(mask_attrs.shrink_axis),
+      ConvertMaskToAxes(mask_attrs.ellipsis));
+
+  // To handle cases like x[2:2], where shape(x) = [1],
+  // TF returns shape = [0], empty vector
+  // make_slice_plan returns begin=2, end=2, but that is > 1
+  // So must clamp them
+  // Another example:
+  // for dimension 3, Also 2:3:-1 gives 4:4, which will also fail if we try to
+  // construct slice. So must clamp to 2:2 etc
+  auto clamp = [](int64_t x, int64_t min, int64_t max) {
+    return x > max ? max : (x < min ? min : x);
+  };
+  for (int i = 0; i < shape.size(); i++) {
+    slice_plan.begins[i] = clamp(slice_plan.begins[i], 0, shape[i]);
+    slice_plan.ends[i] = clamp(slice_plan.ends[i], 0, shape[i]);
+  }
+
+  NGRAPH_VLOG(5) << "Slice_plan: begin: " << ng::join(slice_plan.begins)
+                 << ", end: " << ng::join(slice_plan.ends)
+                 << ", stride: " << ng::join(slice_plan.strides)
+                 << ", reshape input shape: " << slice_plan.reshape_in_shape
+                 << ", reshape output shape: " << slice_plan.reshape_out_shape
+                 << ", reverse axis: " << slice_plan.reverse_axes;
+
+  return Status::OK();
```

Comment:
This function returns only Status::OK() at the end. So maybe we can make it void or just make it return slice_plan directly

Reply:
I wondered about this and should probably ask first. Some of the functions in ngraph_builder.cc return status::OK() and pass returns as reference in the functions, such as `GetInputNodes`, `GetStaticInputVector`, etc.., while others don't. I wonder what's the standard for bridge to decide which function needs to return a status while others could return a value.

Reply:
Sometimes we may not necesarily call any TF function. We may just do some normal cpp op. But if we expect it might throw an error we can return a TF status:

```
Status my_divide(int y, node* n, int& z) {
int x;
TF_RETURN_IF_ERROR(n->get_attribute("foo", &x));
// If "foo" doesnt exist return teh status TF function get_attribute throws

// Got x.
if (y == 0){
  return errors::Internal("Bad denominator");
  // we generate an error message. This error message is not coming from a TF function call
} else {
  z = x/y;
}

return Status::OK();
}
```